### ISSUE TITLE
Merge connected lines with Line Orientation Dependent position label error

### ISSUE THREAD

**Issue by pjshelton** [qgis/QGIS#38205]
#28628 

With the following layer.

geom, volume
Linestring(0 0, 10 0), 20
Linestring(10 0, 0 0), 21
Linestring(20 0, 10 0), 20
Linestring(10 0, 20 0), 23
Linestring(10 0, 10 10), 2
Linestring(10 10, 10 0), 6

set the label _value_ to volume, check _Label placement_ to 'Parallel', and _Allowed positions_ to 'Left of line' and 'Line orientation dependent position'.
Under _Rendering_ check 'Merge connected lines to avoid duplicate labels.  

Volume represents road traffic and is the number of cars going from start to end on each line.

What has happened is that the Merge has now merged the west bound and east bound traffic as they are both 20, where as, because they are going in opposite directions, and therefore are meant to be rendered on different sides of the line, I don't want them merged. 

Is there anyway to add knowledge about which side of the line a label is expected to be placed (when the line dependent checkbox has also been ticked) to the merge connected algorithm?

I know it happens in all versions from at least 3.4.2

**Comment by gioman** [comment]
> Is there anyway to add knowledge about which side of the line a label is expected to be placed (when the line dependent checkbox has also been ticked) to the merge connected algorithm?

@pjshelton is this a question or a bug report? Your use case seems to be a very specific one, where you use two identical (overlapping) lines to represent opposite directions. It seems that the "merge" option does its job for any line that share at least 1 point to another one (as expected for the general use case).

**Comment by pjshelton** [comment]
Yes it is a very specific use case. But I submit that Traffic Engineers and Transport Modellers world over use this method of representing roads. And it could also be useful for any one doing movement mapping where the path in and the path out use the same infrastructure. 
Is it a bug report? At the very least it is a feature request. I used a bug report form because when I use these two options, 'Line direction dependent labeling' and 'Merge' **together**,  QGIS does the _wrong_ thing.  Independently, each of these options are working perfectly. Used together, they break expectation. 
Is it a question? Yes.  The solution that I am suggesting is that when **both** options are active together, the "merge" algorithm should add a restriction to the condition, that not only that it the line has to share at least one point, but that they both go in the same direction.  and if they share only one point that that point is the end on one line and the start of the other.  As I am not set up to create a patch my self, my question is, a) is this an **acceptable** solution for what should happen when **both** options are checked.  And b) if it is, can I ask that someone who knows the code please add the restriction and check.

**Comment by pjshelton** [comment]
Yes for a given "road" both lines are within the other, but they are not exact copies as each one lists the vertices in opposite directions.

**Comment by gioman** [comment]
> Used together, they break expectation

they don't if you think about the general case, two lines that share the same attribute values and that are connected by 1 point. I think that this is more suited for a feature request.

**Comment by pjshelton** [comment]
Maybe.
I think that if you have chosen to use the **Line Orientation Dependent Position** option, you are making the statement that the direction matter. Therefore for the merge to ignore it is wrong

**Comment by pjshelton** [comment]
If you don't care about direction, then why use the **Line Orientation Dependent Position**?  It is only when I try to use both of these together, because I care about the direction of the line, and I want to reduce label clutter, that I have an issue.

**Comment by roya0045** [comment]
couldn't you just offset the line by x meter to the left of right. This way you would have two lines for each traffic direction as you want to represent traffic flow rather than the full road.

**Comment by pjshelton** [comment]
I do that as well, (using the geometry generator for the symbology) but at the scale I need to display I cannot separate them far enough for the difference to be useful without also using the **Line Orientation Dependent Position** option for the labeling.  The encoded geometry (as stored in the layer) is nominally the centerline of the road
![image](https://user-images.githubusercontent.com/39154868/89836283-be7c6200-dba9-11ea-9dac-8acc975cebfd.png)
.

**Comment by nyalldawson** [comment]
My 2c: I think it's a valid feature request, but isn't a bug. It's also a relatively niche feature request, so I'd hint to the submitter that if you want this resolved in a timely manner the best approach would be to raise some funding for its development.

@pjshelton I've gotta say -- I really like the look of that map! It's simple, but the symbology and label choices are extremely refined. Lovely work!

**Comment by pjshelton** [comment]
I am happy with it being a feature request.  (I don't have the required permissions to change the labels.)

 @nyalldawson Thank for the comment on my map making skills.

**Comment by pjshelton** [comment]
@roya0045. Two of the four modelling packages that I have access to model roads the way I have shown, and one models them as single object with directional attributes labeled with AB_ and BA_,  And I am using GIS as a visualisation tool, so  I can not change how the links are coded.  
The one package that does use separated lines like you are suggesting also leaves gaps between each section because the intersections take up real space, and the movement indicators (turns) are not part of the layer that contains the link volumes, so using the merge option is not an option.

**Comment by pjshelton** [comment]
Giving this more thought, and referencing the network supplied in #28628 it would be nice if there was also a way to list exceptions.  links that touch, have the same attribute but you don't want merged.
